include/tvm/topi/transform.h

```diff
@@ -2035,6 +2034,73 @@ inline Tensor adv_index(const Tensor& data, const Array<Tensor>& indices,
       name, tag);
 }
 
+namespace relax {
+// relax dynamic slice
+inline te::Tensor dynamic_strided_slice(const te::Tensor& x, const te::Tensor& begin,
+                                        const te::Tensor& end, const te::Tensor& strides,
+                                        Array<PrimExpr> output_shape,
+                                        std::string name = "T_strided_slice_dynamic",
+                                        std::string tag = kInjective) {
```

Comment:
Why can't we use the existing definition of `dynamic_strided_slice`? https://github.com/apache/tvm/blob/4e07a8ed6687a08b6b27db21af019a5a179b9ee1/include/tvm/topi/transform.h#L648-L651

Reply:
sorry maybe I'm confused about the existing implementation of `dynamic_strided_slice` and its relation with `strided_slice` shape func. If `dynamic_strided_slice` can calculate the right output shape from its dynamic input, why do we need a shape func for it at all...

Reply:
No worries! :)

Reply:
so do you have an answer to my original question? Why can't we simply create the output shape by

```
  Array<PrimExpr> out_shape;

  for (size_t i = 0; i < num_slice_axes; ++i) {
    auto d = indexdiv(end[i] - begin[i], strides[i]);
    if (d->IsInstance<tvm::IntImmNode>()) {
      // Preserve static dimension if possible
      out_shape.push_back(d);
    } else {
      out_shape.push_back(tvm::tir::Var("dim"));
    }
  }
```
?

Sorry if I'm missing something obvious.

Reply:
My apology. It seems like I misread your comment. 

The problem was how to insert match cast logic and bind those variables. I couldn't see a way to insert them in TOPI side.  So I did it in the legalizer and it became explicit and straightforward. 
https://github.com/apache/tvm/blob/4901e3db16673ee48776cde3bff2024efc2d3a39/python/tvm/relax/transform/legalize_ops/index.py#L77-L88

Reply:
This, again, sounds odd to me to talk about legalizer vs topi distinction. Since `dynamic_strided_slice` and the shape func is only meant to be used by the legalizer, why not just define them here? There is no need to be concerned about "TOPI side".